src/emerge.cpp

```diff
@@ -673,9 +679,38 @@ void *EmergeThread::run()
 			continue;
 
 		bool allow_gen = bedata.flags & BLOCK_EMERGE_ALLOW_GEN;
+		bool disk_tried = false;
 		EMERGE_DBG_OUT("pos=" << pos << " allow_gen=" << allow_gen);
 
-		action = getBlockOrStartGen(pos, allow_gen, &block, &bmdata);
+retry:
+		action = getBlockOrStartGen(pos, !disk_tried, allow_gen, &block, &bmdata);
+
+		/* Try to load it */
+		if (action == EMERGE_FROM_DISK) {
+			auto &m_db = *m_emerge->m_db;
+			std::string data;
+			{
+				ScopeProfiler sp(g_profiler, "EmergeThread: load block - async (sum)");
+				MutexAutoLock dblock(m_db.mutex);
+				m_db.loadBlock(pos, data);
+			}
+			if (!data.empty()) {
+				MutexAutoLock envlock(m_server->m_env_mutex);
+				m_map->loadBlock(data, pos);
+			}
+			// We took our shot, check again:
+			disk_tried = true;
+			goto retry;
```

Comment:
Would be easier to understand control flow, imo:
```suggestion
			// We took our shot, check again:
			action = getBlockOrStartGen(pos, false, allow_gen, &block, &bmdata);
			if (action == EMERGE_FROM_MEMORY) {
				// great, but correct the status in case we just loaded it
				action = EMERGE_FROM_DISK;
			}
```